f2py (1.20) drops definition`typedef signed char signed_char;` for callbacks in some circumstances

<!-- Please describe the issue in detail here, and fill in the fields below -->

### Reproducing code example:

<!-- A short code example that reproduces the problem/missing feature. It should be
self-contained, i.e., possible to run as-is via 'python myproblem.py' -->

Unfortunately this is part of a larger project.
And may be tricky to reproduce.
Depending on other definition of callback functions, I get the error message below for the *second* routine.

this causes the error:

```python
subroutine pyexit(code)

  implicit none
  integer(kind=4), intent(IN):: code

  !f2py    integer intent(in): code
  !f2py    intent(callback, hide) endkepler()
  external endkepler

  call endkepler()

end subroutine pyexit

!=======================================================================

subroutine pygets(ttymsg)

  implicit none
  character*(*), intent(inout) :: ttymsg

  integer, parameter :: n = 132
  integer :: i
  integer(kind=1), dimension(n) :: data

  !f2py    intent(callback, hide) ttykepler(data)
  external ttykepler

  call ttykepler(data)

  do i = 1, min(n, len(ttymsg))
     ttymsg(i:i) = char(data(i))
  end do

end subroutine pygets
```
but if I change the *first* routine to

```python
subroutine pyexit(code)

  implicit none
  integer(kind=4), intent(IN):: code

  !f2py    integer intent(in): code
  !f2py    intent(callback, hide) endkepler(code)
  external endkepler

  call endkepler(code)
end subroutine pyexit
```
then the compilation error in the *second* routine does not occur.



### Error message:

```
...
/tmp/tmp2skuwcza/src.linux-x86_64-3.9/_kepler_NBURN_8192_JMZ_1983_FULDAT_fuldat1_f_CPU_Intel_R_Xeon_R_W_2145_CPU_3_70GHzmodule.c:29:63: error: unknown type name ‘signed_char’; did you mean ‘signed char’?
   29 | typedef void(*cb_ttykepler_in_pygets__user__routines_typedef)(signed_char *);
...
```

### NumPy/Python version information:

<!-- Output from 'import sys, numpy; print(numpy.__version__, sys.version)' -->

1.20.0 3.9.1 (default, Dec 16 2020, 01:02:01) 
[GCC 10.2.1 20201125 (Red Hat 10.2.1-9)]

@pearu Thanks for looking at this.

Yes, I realised the missing `::`.  My apologies.
But it does not change the outcome for me.  Still getting same error message on compile.

I was trying to grep the f2py code, but could not spot an obvious place where there might be a confusion between `signed_char` and `signed char`.  Even if the statement in `!f2py` had been wrong, it still should not produce C code with wrong data types, I would say.  So something deeper must be going wrong.

Yes, suspected this will not be easy to reproduce, due to the odd interaction - the compile error is in a different subroutine than the one that was changed.  Which seems odd enough to me.

```python
subroutine pyexit(code)

  implicit none
  integer(kind=4), intent(IN) :: code

  !f2py    integer intent(in) :: code
  !f2py    intent(callback, hide) :: endkepler()
  external endkepler

  call endkepler()

end subroutine pyexit


!=======================================================================

subroutine pygets(ttymsg)

  implicit none
  character*(*), intent(inout) :: ttymsg

  integer, parameter :: n = 132
  integer :: i
  integer(kind=1), dimension(n) :: data

  !f2py    intent(callback, hide) :: ttykepler(data)
  external ttykepler

  call ttykepler(data)

  do i = 1, min(n, len(ttymsg))
     ttymsg(i:i) = char(data(i))
  end do
end subroutine pygets
```
still getting 

```
....
/tmp/tmprjo5dtlj/src.linux-x86_64-3.9/_kepler_NBURN_8192_JMZ_1983_FULDAT_fuldat1_f_CPU_Intel_R_Xeon_R_W_2145_CPU_3_70GHzmodule.c: At top level:
/tmp/tmprjo5dtlj/src.linux-x86_64-3.9/_kepler_NBURN_8192_JMZ_1983_FULDAT_fuldat1_f_CPU_Intel_R_Xeon_R_W_2145_CPU_3_70GHzmodule.c:29:63: error: unknown type name ‘signed_char’; did you mean ‘signed char’?
   29 | typedef void(*cb_ttykepler_in_pygets__user__routines_typedef)(signed_char *);
      |                                                               ^~~~~~~~~~~
      |                                                               signed char

...

/tmp/tmprjo5dtlj/src.linux-x86_64-3.9/_kepler_NBURN_8192_JMZ_1983_FULDAT_fuldat1_f_CPU_Intel_R_Xeon_R_W_2145_CPU_3_70GHzmodule.c: At top level:
/tmp/tmprjo5dtlj/src.linux-x86_64-3.9/_kepler_NBURN_8192_JMZ_1983_FULDAT_fuldat1_f_CPU_Intel_R_Xeon_R_W_2145_CPU_3_70GHzmodule.c:831:42: error: unknown type name ‘signed_char’; did you mean ‘signed char’?
      |                                          ^~~~~~~~~~~
      |                                          signed char
...
```

actually, no, for me if I just have
```fortran
subroutine pyexit(code)

  implicit none
  integer(kind=4), intent(IN):: code

  !f2py    integer intent(in) :: code
  !f2py    intent(callback, hide) :: endkepler()
  external endkepler

  call endkepler()

end subroutine pyexit
!=======================================================================

subroutine pygets(ttymsg)

  implicit none
  character*(*), intent(out) :: ttymsg

  integer, parameter :: n = 132
  integer :: i
  integer(kind=1), dimension(n) :: data

  !f2py    character*(*), intent(out) :: ttymsg
  !f2py    intent(callback, hide) :: ttykepler(data)
  external ttykepler

  call ttykepler(data)

  do i = 1, min(n, len(ttymsg))
     ttymsg(i:i) = char(data(i))
  end do

end subroutine pygets
```
and all else commented out, I already get the error message
```shell
/home/alex/python/source/kepler/code/_kepler_build/src.linux-x86_64-3.9/_kepler_NBURN_8192_JMZ_1983_FULDAT_fuldat1_f_CPU_Intel_R_Xeon_R_W_2145_CPU_3_70GHzmodule.c: At top level:
/home/alex/python/source/kepler/code/_kepler_build/src.linux-x86_64-3.9/_kepler_NBURN_8192_JMZ_1983_FULDAT_fuldat1_f_CPU_Intel_R_Xeon_R_W_2145_CPU_3_70GHzmodule.c:622:42: error: unknown type name ‘signed_char’; did you mean ‘signed char’?
  622 | extern void F_FUNC(ttykepler,TTYKEPLER) (signed_char *data) {
      |                                          ^~~~~~~~~~~
      |                                          signed char
```
Is it conceivable that commented-out parts cause issues?

--- a/numpy/f2py/cb_rules.py
+++ b/numpy/f2py/cb_rules.py
@@ -342,6 +342,7 @@ cb_arg_rules = [
             isarray: '#ctype# *',
             isstring: '#ctype#'
         },
+        'need': {l_or(isscalar, isarray, isstring): '#ctype#'},
         # untested with multiple args
         'strarglens': {isstring: ',int #varname_i#_cb_len'},
         'strarglens_td': {isstring: ',int'},  # untested with multiple args

--- a/numpy/f2py/tests/test_callback.py
+++ b/numpy/f2py/tests/test_callback.py
@@ -236,3 +236,43 @@ class TestF90Callback(util.F2PyTest):
         y = np.array([1, 2, 3], dtype=np.int64)
         r = self.module.gh17797(incr, y)
         assert r == 123 + 1 + 2 + 3
+
+
+class TestGH18335(util.F2PyTest):
+    """The reproduction of the reported issue requires specific input that
+    extensions may break the issue conditions, so the reproducer is
+    implemented as a separate test class. Do not extend this test with
+    other tests!
+    """
+
+    suffix = '.f90'
+
+    code = textwrap.dedent(
+        """
+        ! When gh18335_workaround is defined as an extension,
+        ! the issue cannot be reproduced.
+        !subroutine gh18335_workaround(f, y)
+        !  implicit none
+        !  external f
+        !  integer(kind=1) :: y(1)
+        !  call f(y)
+        !end subroutine gh18335_workaround
+
+        function gh18335(f) result (r)
+          implicit none
+          external f
+          integer(kind=1) :: y(1), r
+          y(1) = 123
+          call f(y)
+          r = y(1)
+        end function gh18335
+        """)
+
+    def test_gh18335(self):
+
+        def foo(x):
+            x[0] += 1
+
+        y = np.array([1, 2, 3], dtype=np.int8)
+        r = self.module.gh18335(foo)
+        assert r == 123 + 1

